core/src/main/scala/kafka/log/remote/RemoteLogIndexEntry.scala

```diff
@@ -0,0 +1,229 @@
+          val entryLength = magicAndEntryLengthBuffer.getShort()
+          val valueBuffer = ByteBuffer.allocate(entryLength)
+          Utils.readFully(is, valueBuffer)
+          valueBuffer.flip()
+
+          try {
+            val crc = valueBuffer.getInt()
+            val firstOffset = valueBuffer.getLong()
+            val lastOffset = valueBuffer.getLong()
+            val firstTimestamp = valueBuffer.getLong()
+            val lastTimestamp = valueBuffer.getLong()
+            val dataLength = valueBuffer.getInt()
+            val rdiLength = valueBuffer.getShort()
+            val rdi = Array.ofDim[Byte](rdiLength)
+            valueBuffer.get(rdi)
+            Some(RemoteLogIndexEntry(magic, crc, firstOffset, lastOffset, firstTimestamp, lastTimestamp, dataLength, rdi))
+          } catch {
+            case e: BufferUnderflowException =>
+              // TODO decide how to deal with incomplete records
+              throw new KafkaException(e)
+          }
+        case _ =>
+          // TODO: Throw Custom Exceptions?
+          throw new RuntimeException("magic version " + magic + " is not supported")
+      }
+    }
+  }
+
+  def parseEntry(ch: FileChannel, position: Long): Option[RemoteLogIndexEntry] = {
+    val magicBuffer = ByteBuffer.allocate(2)
+    val readCt = ch.read(magicBuffer, position)
+    if (readCt > 0) {
+      magicBuffer.flip()
+      val magic = magicBuffer.getShort
+      magic match {
+        case 0 =>
+          val valBuffer = ByteBuffer.allocate(4)
+          val nextPos = position + 2
+          ch.read(valBuffer, nextPos)
+          valBuffer.flip()
+          val length = valBuffer.getShort
+
+          val valueBuffer = ByteBuffer.allocate(length)
+          ch.read(valueBuffer, nextPos + 2)
+          valueBuffer.flip()
+
+          val crc = valueBuffer.getInt
+          val firstOffset = valueBuffer.getLong
+          val lastOffset = valueBuffer.getLong
+          val firstTimestamp = valueBuffer.getLong
+          val lastTimestamp = valueBuffer.getLong
+          val dataLength = valueBuffer.getInt
+          val rdiLength = valueBuffer.getShort
+          val rdiBuffer = ByteBuffer.allocate(rdiLength)
+          valueBuffer.get(rdiBuffer.array())
+          Some(RemoteLogIndexEntry(magic, crc, firstOffset, lastOffset, firstTimestamp, lastTimestamp, dataLength,
+            rdiBuffer.array()))
+        case _ =>
+          // TODO: Throw Custom Exceptions?
+          throw new RuntimeException("magic version " + magic + " is not supported")
+      }
+    } else {
+      debug(s"Reached limit of the file for channel:$ch for position: $position")
+      None
+    }
+  }
+
+  def readAll(ch: FileChannel): Seq[RemoteLogIndexEntry] = {
+    var index = Seq[RemoteLogIndexEntry]()
+
+    var pos = 0L;
+    while (pos < ch.size()) {
+      val entryOption = parseEntry(ch, pos.toInt)
+      val entry = entryOption.getOrElse(throw new KafkaException(s"Entry could not be found at position: $pos"))
```

Comment:
Kafka uses `CorruptIndexException` for sanity check of offset indexes. Maybe something following the same pattern here e.g. `CorruptRemoteIndexException`?